### DISCUSSION TITLE
How to get the color of bitmap pixel?

### DISCUSSION THREAD

**Discussion by romanIlyushin-bc** [visgl/deck.gl#5241] [Q&A]
[getPickingInfo()](https://deck.gl/docs/api-reference/core/layer#decodepickingcolor) is returning the color of layer. How can I get the color of pixel on BitmapLayer?

**Comment by kylebarron** [comment]
I don't believe it's currently possible to do this. The picking engine is focused on picking vector objects, so an entire bitmap is treated as a single object. I think you'd have to customize the picking shaders to pick within the bitmap.

**Comment by sraimund** [comment]
As a workaround, you can render the bitmap into a separate canvas:

```js
const canvas = document.createElement("canvas");
const canvasContext = canvas.getContext("2d");

const layer = new BitmapLayer({
    id: "BitmapLayer",
    bounds: [-122.519, 37.7045, -122.355, 37.829],
    opacity: 0.5, //the retrieved color stays the same for varying opacities
    image: "https://raw.githubusercontent.com/visgl/deck.gl-data/master/website/sf-districts.png",
    pickable: true,
    onClick: ({bitmap, layer}) => {
        if (bitmap) {
            const x = bitmap.pixel[0];
            const y = bitmap.pixel[1];

            canvas.width = bitmap.size.width;
            canvas.height = bitmap.size.height;
            canvasContext.drawImage(layer.props.image.props.data, 0, 0);

            const pixelColor = canvasContext.getImageData(x, y, 1, 1).data;
            console.log('Color at picked pixel:', pixelColor);
        }
    },
});
```

It does not seem to be possible currently to get the pixel value directly from the ImageBitmap (see [stackoverflow](https://stackoverflow.com/questions/68102995/is-it-possible-to-get-pixels-from-imagebitmap-web-object-without-a-canvas)).

**Comment by lawrencenika** [comment]
is there still no method of getting a pixel value from bitmaplayer?

**Reply by kylebarron** [reply]
https://deck.gl/docs/api-reference/layers/bitmap-layer#pixel-picking

**Reply by lawrencenika** [reply]
thanks, but 2 questions:
1. I see that `readPixelsToArrayWebGL` is marked as deprecated here. Is it recommended in that case?
2. See screenshot, as I tried this piece of code in getToolTip() of DeckGL instead of OnClick() of BitmapLayer it failed, what is the diff?
![image](https://github.com/user-attachments/assets/577c835a-99ee-46dd-b8f1-5125ef9a2a40)

**Reply by kylebarron** [reply]
I don't know, I haven't used it myself.

**Reply by ibgreen** [reply]
- readPixelsToArrayWebGL  is marked deprecated because it won't work in WebGPU. It will become an issue in deck.gl 9.2 but you are best off ignoring that for now.
- What is the type of `pickinginfo.layer.props.mage`? Is it a `Texture`? if not that could be where the undefined reference is coming from...

**Reply by mohadib** [reply]
> * readPixelsToArrayWebGL  is marked deprecated because it won't work in WebGPU. It will become an issue in deck.gl 9.2 but you are best off ignoring that for now.

Do you have any information on what will replace readPixelsToArrayWebGL? Thanks.
